### PR TITLE
Sync to nodejs/master (2016-05-05)

### PR DESCRIPTION
**Pull request by kunalspathak**
##### Checklist

- [x] tests and code linting passes
- [x] a test and/or benchmark is included
- [x] the commit message follows commit guidelines
##### Affected core subsystem(s)

deps
##### Description of change

- Synced nodejs/master as of 2016-05-05.
- Added shim for Proxy class
- Fixed `TypedArray::Length` API in shim

### CODE REVIEW

**Comment by kunalspathak** [comment]
CC: @nodejs/node-chakracore

File: test/parallel/test-process-versions.js [hunk 1]
```diff
@@ -10,7 +10,7 @@ if (common.hasCrypto) {
   expected_keys.push('openssl');
 }
 
-if (!common.isChakraEngine && typeof Intl !== 'undefined') {
+if (typeof Intl !== 'undefined') {
```
Comment (jianchun) [inline test/parallel/test-process-versions.js:13]: Why this works now?

**Comment by jianchun** [comment]
👍

File: test/parallel/test-process-versions.js (hunk 1 above)
Comment (kunalspathak) [inline test/parallel/test-process-versions.js:13]: Because 03a8637f7934b12516d36ea3c5c9fb2443bdfb8d made Intl support enabled by default.

**Pull request merged**